### PR TITLE
Merging sending of mb and forward txn to lookup with DataSender

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

- Applied DataSender for sending forward txn to lookup
- Combined the sending of the microblock from shard to sending forward txn
-- including modifying Protobuf and modifying ForwardTxnEntry => MBnForwardTxnEntry
- Added checker when lookup received forwarded txn and microblock

- A minor fix on the TxnOrderVerifier is included here also

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] small-scale cloud test (200 aws nodes test)

### CODE REVIEW

File: src/libData/AccountData/TransactionReceipt.h [hunk 1]
```diff
@@ -84,28 +84,37 @@ class TransactionWithReceipt : public SerializableDataBlock {
     return m_tranReceipt;
   }
 
+  static TxnHash ComputeTransactionReceiptsHash(
+      const std::vector<TransactionWithReceipt>& txrs) {
+    if (txrs.empty()) {
+      LOG_GENERAL(INFO, "TxnOrder is empty");
```
Comment (ansnunez) [inline src/libData/AccountData/TransactionReceipt.h]: Change `TxnOrder` to `txrs`

File: src/libNetwork/DataSender.cpp [hunk 2]
```diff
@@ -237,16 +237,18 @@ bool DataSender::SendDataToOthers(
     unsigned int my_shards_lo = 0;
     unsigned int my_shards_hi = 0;
 
-    DetermineShardToSendDataTo(my_cluster_num, my_shards_lo, my_shards_hi,
-                               shards, tmpCommittee, indexB2);
+    if (!shards.empty()) {
```
Comment (ansnunez) [inline src/libNetwork/DataSender.cpp]: Should we move the check earlier (e.g., start of this function)?

Comment (kaikawaliu) [reply src/libNetwork/DataSender.cpp]: if shards is empty then only send to lookup

File: src/libNode/FinalBlockProcessing.cpp [hunk 3]
```diff
@@ -487,48 +425,76 @@ void Node::CallActOnFinalblock() {
...
 
-  LOG_GENERAL(INFO, "[SendMB]"
+  // Transaction body sharing
+  mb_txns_message = {MessageType::NODE,
+                     NodeInstructionType::MBNFORWARDTRANSACTION};
+
+  if (!Messenger::SetNodeMBnForwardTransaction(
+          mb_txns_message, MessageOffset::BODY, *m_microblock, txns_to_send)) {
+    LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
+              "Messenger::SetNodeForwardTransaction failed.");
```
Comment (ansnunez) [inline src/libNode/FinalBlockProcessing.cpp]: Change `SetNodeForwardTransaction` to `SetNodeMBnForwardTransaction`

File: src/libNode/FinalBlockProcessing.cpp [hunk 4]
```diff
@@ -958,51 +918,79 @@ void Node::DeleteEntryFromFwdingAssgnAndMissingBodyCountMap(
...
 
   LOG_MARKER();
 
-  ForwardedTxnEntry entry;
+  MBnForwardedTxnEntry entry;
 
-  if (!Messenger::GetNodeForwardTransaction(message, cur_offset, entry)) {
+  if (!Messenger::GetNodeMBnForwardTransaction(message, cur_offset, entry)) {
     LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
               "Messenger::GetNodeForwardTransaction failed.");
```
Comment (ansnunez) [inline src/libNode/FinalBlockProcessing.cpp]: Change `GetNodeForwardTransaction` to `GetNodeMBnForwardTransaction`

File: src/libData/AccountData/MBnForwardedTxnEntry.h [hunk 5]
```diff
@@ -17,27 +17,28 @@
...
-  friend std::ostream& operator<<(std::ostream& os, const ForwardedTxnEntry& t);
+  friend std::ostream& operator<<(std::ostream& os,
+                                  const MBnForwardedTxnEntry& t);
 };
 
-inline std::ostream& operator<<(std::ostream& os, const ForwardedTxnEntry& t) {
-  os << "<ForwardedTxnEntry>" << std::endl
-     << "m_blockNum : " << t.m_blockNum << std::endl
-     << "m_hash : " << t.m_hash.hex();
+inline std::ostream& operator<<(std::ostream& os,
```
Comment (KaustubhShamshery) [inline src/libData/AccountData/MBnForwardedTxnEntry.h]: can include mb hash?

Comment (kaikawaliu) [reply src/libData/AccountData/MBnForwardedTxnEntry.h]: I wish someday we can have a printing operator for all the block data, then no need to manually log these every time, will put an issue for that

**Pull request merged**